python/pyspark/sql/pandas/conversion.py

```diff
@@ -0,0 +1,431 @@
+                        "computation.\n  %s" % _exception_message(e))
+                    warnings.warn(msg)
+                    raise
+
+        # Below is toPandas without Arrow optimization.
+        pdf = pd.DataFrame.from_records(self.collect(), columns=self.columns)
+
+        dtype = {}
+        for field in self.schema:
+            pandas_type = PandasConversionMixin._to_corrected_pandas_type(field.dataType)
+            # SPARK-21766: if an integer field is nullable and has null values, it can be
+            # inferred by pandas as float column. Once we convert the column with NaN back
+            # to integer type e.g., np.int16, we will hit exception. So we use the inferred
+            # float type, not the corrected type from the schema in this case.
+            if pandas_type is not None and \
+                not(isinstance(field.dataType, IntegralType) and field.nullable and
+                    pdf[field.name].isnull().any()):
+                dtype[field.name] = pandas_type
+            # Ensure we fall back to nullable numpy types, even when whole column is null:
+            if isinstance(field.dataType, IntegralType) and pdf[field.name].isnull().any():
+                dtype[field.name] = np.float64
+            if isinstance(field.dataType, BooleanType) and pdf[field.name].isnull().any():
+                dtype[field.name] = np.object
+
+        for f, t in dtype.items():
+            pdf[f] = pdf[f].astype(t, copy=False)
+
+        if timezone is None:
+            return pdf
+        else:
+            from pyspark.sql.pandas.types import _check_series_convert_timestamps_local_tz
+            for field in self.schema:
+                # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+                if isinstance(field.dataType, TimestampType):
+                    pdf[field.name] = \
+                        _check_series_convert_timestamps_local_tz(pdf[field.name], timezone)
+            return pdf
+
+    @staticmethod
+    def _to_corrected_pandas_type(dt):
+        """
+        When converting Spark SQL records to Pandas :class:`DataFrame`, the inferred data type
+        may be wrong. This method gets the corrected data type for Pandas if that type may be
+        inferred incorrectly.
+        """
+        import numpy as np
+        if type(dt) == ByteType:
+            return np.int8
+        elif type(dt) == ShortType:
+            return np.int16
+        elif type(dt) == IntegerType:
+            return np.int32
+        elif type(dt) == LongType:
+            return np.int64
+        elif type(dt) == FloatType:
+            return np.float32
+        elif type(dt) == DoubleType:
+            return np.float64
+        elif type(dt) == BooleanType:
+            return np.bool
+        elif type(dt) == TimestampType:
+            return np.datetime64
+        else:
+            return None
+
+    def _collect_as_arrow(self):
```

Comment:
Here, I piggyback a nit fix. It was `_collectAsArrow`. Unless it's to match with Scala side, it has to follow snake_naming per PEP 8.